File: skills/developing-kafka-java-client/references/JsonSchemaConsumer.java
package com.example.kafka;

import io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer;
import io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializerConfig;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.errors.WakeupException;
import org.apache.kafka.common.serialization.StringDeserializer;

import java.time.Duration;
import java.util.List;
import java.util.Properties;

/**
 * JSON Schema consumer. Deserializes into the {@link JsonSchemaProducer.Transaction}
 * POJO via KafkaJsonSchemaDeserializer. Uses the wakeup + close() shutdown pattern.
 */
public class JsonSchemaConsumer {

    public static void consume(Consumer<String, JsonSchemaProducer.Transaction> consumer, String topic) {
        consumer.subscribe(List.of(topic));
        System.out.println("Listening on topic: " + topic);
        Runtime.getRuntime().addShutdownHook(new Thread(consumer::wakeup));

        try {
            while (true) {
                ConsumerRecords<String, JsonSchemaProducer.Transaction> records =
                        consumer.poll(Duration.ofMillis(1000));
                for (ConsumerRecord<String, JsonSchemaProducer.Transaction> record : records) {
                    System.out.printf("key=%s value=%s%n", record.key(), record.value());
                }
            }
        } catch (WakeupException e) {
            // Expected on shutdown.
        } finally {
            consumer.close();
            System.out.println("Consumer closed");
        }
    }

    public static void main(String[] args) {
        KafkaConfig.Env env = KafkaConfig.loadEnv();
        String topic = env.get("TOPIC", "demo-topic");

        Properties props = KafkaConfig.baseProperties(env);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, env.get("GROUP_ID", "java-consumer-group"));
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaJsonSchemaDeserializer.class.getName());
        // Tell the deserializer which POJO to bind the JSON value to.
        props.put(KafkaJsonSchemaDeserializerConfig.JSON_VALUE_TYPE, JsonSchemaProducer.Transaction.class.getName());

        if (!KafkaConfig.verifyKafkaSetup(props, topic)) {
            throw new IllegalStateException("Failed to verify Kafka setup");
        }
        String srUrl = env.get("SCHEMA_REGISTRY_URL");
        if (!KafkaConfig.verifySchemaRegistry(srUrl, env.get("SR_API_KEY"), env.get("SR_API_SECRET"))) {
            throw new IllegalStateException("Failed to connect to Schema Registry");
        }

        Consumer<String, JsonSchemaProducer.Transaction> consumer = new KafkaConsumer<>(props);
        consume(consumer, topic);
    }
}

File: skills/developing-kafka-dotnet-client/references/JsonSchemaConsumer.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Confluent.Kafka;
using Confluent.Kafka.SyncOverAsync;
using Confluent.SchemaRegistry;
using Confluent.SchemaRegistry.Serdes;

namespace ExampleKafka
{
    /// <summary>
    /// Deserializes via Schema Registry -- no raw string/JSON fallback. Uses a
    /// CancellationToken for graceful shutdown (the .NET idiom): Consume(token)
    /// throws OperationCanceledException when the token is cancelled, instead of
    /// Java's consumer.wakeup() + WakeupException. Commits offsets explicitly
    /// after processing (EnableAutoCommit = false).
    /// </summary>
    public static class JsonSchemaConsumer
    {
        public static async Task Main()
        {
            var env = KafkaConfig.LoadEnv();
            var consumerConfig = KafkaConfig.BaseConsumerConfig(env);
            // KIP-848 next-gen rebalance protocol (Kafka 4.0+ clients and brokers).
            // Eliminates stop-the-world rebalances. Falls back to Classic if unset.
            consumerConfig.GroupProtocol = GroupProtocol.Consumer;
            var topic = KafkaConfig.Get(env, "TOPIC", "demo-topic");

            var adminConfig = KafkaConfig.BaseAdminConfig(env);
            if (!KafkaConfig.VerifyKafkaSetup(adminConfig, topic))
            {
                throw new InvalidOperationException("Failed to verify Kafka setup");
            }

            var srUrl = KafkaConfig.Get(env, "SCHEMA_REGISTRY_URL");
            var srKey = KafkaConfig.Get(env, "SR_API_KEY");
            var srSecret = KafkaConfig.Get(env, "SR_API_SECRET");
            if (!await KafkaConfig.VerifySchemaRegistryAsync(srUrl, srKey, srSecret))
            {
                throw new InvalidOperationException("Failed to connect to Schema Registry");
            }

            using var schemaRegistry = new CachedSchemaRegistryClient(KafkaConfig.SchemaRegistryConfig(env));
            // JsonDeserializer<T> deserializes the JSON body directly (after skipping the Confluent
            // wire-format header) and does not need an ISchemaRegistryClient -- unlike AvroDeserializer,
            // it never fetches the writer schema from the registry to decode a record.
            var deserializer = new JsonDeserializer<Value>(schemaRegistry).AsSyncOverAsync();

            using var consumer = new ConsumerBuilder<string, Value>(consumerConfig)
                .SetValueDeserializer(deserializer)
                .Build();

            using var cts = new CancellationTokenSource();
            Console.CancelKeyPress += (_, e) =>
            {
                e.Cancel = true; // prevent immediate process termination
                cts.Cancel();
            };

            consumer.Subscribe(topic);
            try
            {
                while (true)
                {
                    try
                    {
                        var record = consumer.Consume(cts.Token);
                        Process(record);
                        consumer.Commit(record);
                    }
                    catch (ConsumeException ex)
                    {
                        Console.Error.WriteLine($"Consume error: {ex.Error.Reason}");
                    }
                }
            }
            catch (OperationCanceledException)
            {
                // Expected on shutdown (Ctrl+C / CancellationToken cancelled) -- fall through.
            }
            finally
            {
                consumer.Close();
            }
        }

        private static void Process(ConsumeResult<string, Value> record)
        {
            Console.WriteLine($"Consumed {record.TopicPartitionOffset}: {record.Message.Key} -> {record.Message.Value.OrderId}");
        }
    }
}
